experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts

```diff
@@ -131,120 +139,131 @@ export const optionalChangeRebaser: FieldChangeRebaser<OptionalChangeset> = {
+			const srcToDst = new ChildChangeMap<
+				[dst: RegisterId, target: "nodeTargeting" | "cellTargeting"]
+			>();
+			const dstToSrc = new ChildChangeMap<RegisterId>();
+			for (const [src, dst, target] of moves) {
+				srcToDst.set(src, [dst, target]);
+				dstToSrc.set(dst, src);
 			}
+			return { srcToDst, dstToSrc };
 		};
 
-		let fieldChange: Mutable<OptionalFieldChange> | undefined;
-		let currentChildNodeChanges: TaggedChange<NodeChangeset>[] = [];
-		let index = 0;
+		let latestReservedDetachId: RegisterId | undefined;
+		let isInputContextEmpty: boolean | undefined;
+
+		const builds: OptionalChangeset["build"] = [];
+		const childChangesByOriginalId = new ChildChangeMap<TaggedChange<NodeChangeset>[]>();
+		// TODO: It might be possible to compose moves in place rather than repeatedly copy.
+		// Additionally, working out a 'register allocation' strategy which enables frequent cancellation of noop moves
+		// for sandwich rebases would help with cloning if in-place proves too difficult
+		const current = getBidirectionalMaps([]);
 		for (const { change, revision } of changes) {
-			const fieldChangeInfo = revisionMetadata.tryGetInfo(
-				revision ?? change.fieldChange?.revision,
-			);
-			const { childChanges } = change;
-			if (childChanges !== undefined) {
-				for (const [childId, childChange] of childChanges) {
-					const taggedChildChange = tagChange(childChange, revision);
-					if (childId === "self") {
-						// childChange refers to the node that existed at the start of `change`,
-						// Thus in the composition, it should be referred to by whatever deletes that node in the future, which is what
-						// currentChildNodeChanges tracks
-						currentChildNodeChanges.push(taggedChildChange);
-					} else {
-						addChildChange(childId, taggedChildChange);
-					}
-				}
+			if (
+				isInputContextEmpty === undefined &&
+				(change.moves.length > 0 || change.reservedDetachId !== undefined)
+			) {
+				// change includes sets to the field, thus the emptiness of it previously can be inferred.
+				// Note that this block isn't entered for child-only changes.
+				isInputContextEmpty = change.reservedDetachId !== undefined;
 			}
+			const withIntention = (id: RegisterId): RegisterId => {
+				if (id === "self") {
+					return id;
+				}
+				const intention = getIntention(id.revision ?? revision, revisionMetadata);
+				return { revision: intention, localId: id.localId };
+			};
 
-			if (change.fieldChange !== undefined) {
-				if (fieldChange === undefined) {
-					fieldChange = {
-						id: change.fieldChange.id,
-						revision: fieldChangeInfo?.revision,
-						wasEmpty: change.fieldChange.wasEmpty,
-					};
-				} else {
-					fieldChange.id = change.fieldChange.id;
-					fieldChange.revision = fieldChangeInfo?.revision;
+			if (change.build !== undefined) {
+				for (const { id, set } of change.build) {
+					builds.push({
+						id: {
+							revision: getIntention(id.revision ?? revision, revisionMetadata),
```

Comment:
Since inverses never produce builds it shoud be sufficient to just do:
```suggestion
							revision: id.revision ?? revision,
```
I'm not really concerned about the perf, I mostly want to keep the use of `revisionMetadata` to a minimum so we can keep track of when and why it is needed.